### ISSUE TITLE
Avro nested

### ISSUE THREAD

**Issue by maqroll** [ClickHouse/ClickHouse#9707]
Avro and AvroConfluent formats (20.3.2.1) are a really nice feature (thanks!!). 

Current implementation (AFAIK) **doesn't allow reading nested structures**.

Most Avro messages in real use-cases (i.e. CDC through Debezium) include nested structures, so feeding those messages into clickhouse will require some kind of pre-processing to flatten them.

Debezium (i.e.) produces (default config) messages of this kind:
```
{
	before: {
		// table fields before
	},
	after: {
		// table fields after
	},
	source: {
		// metadata
	}
}
```

**It would be really nice being able to access nested fields directly like before.fieldA or after.fieldB** (suggestion).

There are a lot of issues/questions related to the ability to read nested structures from Kafka in other formats (#4533, #8577, #4533, #8124, etc...).

Opinions/feedback/alternatives?

**Comment by filimonov** [comment]
> are a really nice feature (thanks!!).

kudos to @oandrew :) I don't have too much experience with Avro, @oandrew what do you think about adding support for nested?

**Comment by oandrew** [comment]
Sounds good to me, I will try to add it sometime soon (probably this/next week).

**Comment by maqroll** [comment]
Inspiration: [catj](https://github.com/soheilpro/catj)

**Comment by oandrew** [comment]
https://github.com/tomnomnom/gron ?

**Comment by oandrew** [comment]
@maqroll 
Could you test if my PR covers your use case?
You can download a build from https://clickhouse-builds.s3.yandex.net/10354/68f2e934439766f235361abd4339164c8b8dd7d0/report.html

**Comment by maqroll** [comment]
@oandrew 

Fails with
```
2020.04.20 10:02:16.486936 [ 53 ] {} <Error> void DB::StorageKafka::threadFunc(): Code: 8, e.displayText() = DB::Exception: Field Envelope.source.name not found in Avro schema, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x105bec00 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x8f0e57d in /usr/bin/clickhouse
2. ? @ 0xdbf6e74 in /usr/bin/clickhouse
3. DB::AvroConfluentRowInputFormat::getOrCreateDeserializer(unsigned int) @ 0xdbf4df8 in /usr/bin/clickhouse
4. DB::AvroConfluentRowInputFormat::readRow(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, DB::RowReadExtension&) @ 0xdbf54f8 in /usr/bin/clickhouse
5. DB::IRowInputFormat::generate() @ 0xdc0c371 in /usr/bin/clickhouse
6. DB::ISource::work() @ 0xdb88d3b in /usr/bin/clickhouse
7. ? @ 0xd83f83d in /usr/bin/clickhouse
8. DB::KafkaBlockInputStream::readImpl() @ 0xd84046b in /usr/bin/clickhouse
9. DB::IBlockInputStream::read() @ 0xcde188f in /usr/bin/clickhouse
10. DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::__1::atomic<bool>*) @ 0xce2065f in /usr/bin/clickhouse
11. DB::StorageKafka::streamToViews() @ 0xd82c98d in /usr/bin/clickhouse
12. DB::StorageKafka::threadFunc() @ 0xd82d2a8 in /usr/bin/clickhouse
13. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0xcf8fe75 in /usr/bin/clickhouse
14. DB::BackgroundSchedulePool::threadFunction() @ 0xcf90492 in /usr/bin/clickhouse
15. ? @ 0xcf905c2 in /usr/bin/clickhouse
16. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x8f33f6b in /usr/bin/clickhouse
17. ? @ 0x8f32453 in /usr/bin/clickhouse
18. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
19. __clone @ 0x12188f in /lib/x86_64-linux-gnu/libc-2.27.so
 (version 20.4.1.3079 (official build))
```

using schema:
```
{
    "subject": "inventory.inventory.products-value",
    "version": 1,
    "id": 12,
    "schema": "{\"type\":\"record\",\"name\":\"Envelope\",\"namespace\":\"inventory.inventory.products\",\"fields\":[{\"name\":\"before\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Value\",\"fields\":[{\"name\":\"id\",\"type\":\"int\"},{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"description\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"weight\",\"type\":[\"null\",\"double\"],\"default\":null}],\"connect.name\":\"inventory.inventory.products.Value\"}],\"default\":null},{\"name\":\"after\",\"type\":[\"null\",\"Value\"],\"default\":null},{\"name\":\"source\",\"type\":{\"type\":\"record\",\"name\":\"Source\",\"namespace\":\"io.debezium.connector.mysql\",\"fields\":[{\"name\":\"version\",\"type\":\"string\"},{\"name\":\"connector\",\"type\":\"string\"},{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"ts_ms\",\"type\":\"long\"},{\"name\":\"snapshot\",\"type\":[{\"type\":\"string\",\"connect.version\":1,\"connect.parameters\":{\"allowed\":\"true,last,false\"},\"connect.default\":\"false\",\"connect.name\":\"io.debezium.data.Enum\"},\"null\"],\"default\":\"false\"},{\"name\":\"db\",\"type\":\"string\"},{\"name\":\"table\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"server_id\",\"type\":\"long\"},{\"name\":\"gtid\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"file\",\"type\":\"string\"},{\"name\":\"pos\",\"type\":\"long\"},{\"name\":\"row\",\"type\":\"int\"},{\"name\":\"thread\",\"type\":[\"null\",\"long\"],\"default\":null},{\"name\":\"query\",\"type\":[\"null\",\"string\"],\"default\":null}],\"connect.name\":\"io.debezium.connector.mysql.Source\"}},{\"name\":\"op\",\"type\":\"string\"},{\"name\":\"ts_ms\",\"type\":[\"null\",\"long\"],\"default\":null}],\"connect.name\":\"inventory.inventory.products.Envelope\"}"
}
```

I tried both Envelope.source.name and source.name on these records (json output from kafkacat):
```
{"before": null, "after": {"Value": {"id": 101, "name": "scooter", "description": {"string": "Small 2-wheel scooter"}, "weight": {"double": 3.1400001049041748}}}, "source": {"version": "1.0.0.Final", "connector": "mysql", "name": "inventory", "ts_ms": 0, "snapshot": {"string": "true"}, "db": "inventory", "table": {"string": "products"}, "server_id": 0, "gtid": null, "file": "mysql-bin.000003", "pos": 154, "row": 0, "thread": null, "query": null}, "op": "c", "ts_ms": {"long": 1587376715994}}
```

using [this setup](https://gist.github.com/maqroll/c3723930c20936c2c27ab9d721da9a5a).

Am I missing something?

**Comment by oandrew** [comment]
@maqroll 
`source.name` should work (tested locally with your schema).

`before.name` (or `after.name`) won't work because `before` field is a `union`.

I am planning to add support for such stuff in another PR.
The column name would then specify the path to a field based on  avro json  encoding (http://avro.apache.org/docs/current/spec.html#json_encoding)

e.g. you would access `before.description` using:
`"before.Value.description" Nullable(String)`
or
`"before.Value.description.string" String`
or
`"before.Value.description.string" Nullable(String)`

**Comment by maqroll** [comment]
@oandrew 
Sorry, I checked against a different build!! 

Being able to access the rest of the fields (union support) would be great.
As I see it all the fields inside before should all be defined as Nullable (even if they are not nullable themselves, id for instance). Something like this:
```
before.Value.id Nullable(int),
before.Value.name Nullable(string),
before.Value.description Nullable(string),
before.Value.weight Nullable(double)
```

``before.Value.description.string`` defined as String (not nullable) doesn't make sense, does it?

Great work!!!

**Comment by oandrew** [comment]
I was thinking about maybe supporting both nullable and non-nullable.
In case of non-nullable it would use default value instead of null e.g. empty string for `String` or 0 for `Int`.

e.g. if you have schema
```
{
  "type": "record",
  "name": "Root",
  "fields": [
    {
      "name": "before",
      "type": [
        "null",
        {
          "type": "record",
          "name": "Value",
          "fields": [
            {
              "name": "field1",
              "type": ["null", "string"]
            },
            {
              "name": "field2",
              "type": ["null", "string", "double"]
            }
          ]
        }
      ]
    }
  ]
}
```

then you would access `field1` with:
`"before.Value.field1" Nullable(String)` - special case support for `union(null, T)`
`"before.Value.field1.string" String` - generic union support + default value
`"before.Value.field1.string" Nullable(String)` - generic union support + nullable

and `field2` with:
`"before.Value.field1.string" String, "before.Value.field1.double" Double` - generic union support + default value
or
`"before.Value.field1.string" Nullable(String), "before.Value.field1.double" Nullable(Double)` - generic union support + default value


To generalize, there would be two ways to access a union:
old: `"a.b.c" Nullable(T)` - if `c`  type is `union(null, T)` or `union(T, null)`
new (generic): `"a.b.c.T" T` / `"a.b.c.T" Nullable(T)`  - if `c` is `union(..,T,..)` i.e. access by branch name

**Comment by maqroll** [comment]
Looks great.

What if `"before.Value.field1.string" String DEFAULT "hello"`?

Would take into account the default value instead of the default default value by type?

**Comment by oandrew** [comment]
@maqroll 
yep, it will use `DEFAULT` if specified

**Comment by oandrew** [comment]
@maqroll 
Could you test again?
Build from here: https://clickhouse-builds.s3.yandex.net/0/436e4e3f5eabbf03e9493c9183cc0c6f8f628882/report.html

**Comment by maqroll** [comment]
@oandrew  

Default values doesn't seem to be working in Kafka tables neither for avro nor avroconfluent format.

If I push a null value defined as union(string,null)  through clickhouse-local and read it using String DEFAULT 'default' I get 'default' (as in some your tests) but I try the same using Kafka tables I get an empty string.

`
cat data.avro | clickhouse local --input-format Avro --output-format CSV -S "\"source.query.string\" String DEFAULT 'a'" -q 'select * from table'
`
returns "a"

but 
`
kafkacat -P -b kafka:9092 -t test -p 0 data.avro
`
followed by
`
CREATE TABLE IF NOT EXISTS kafka_test
(   
    `source.query.string` String default 'a',
) ENGINE = Kafka()
SETTINGS
    kafka_broker_list = 'kafka:9092',
    kafka_topic_list = 'test',
    kafka_group_name = 'ch_test_group',
    kafka_format = 'Avro',
    kafka_num_consumers = 1;
`

reports an empty string.

**Comment by oandrew** [comment]
@maqroll
Looks like Kafka storage does not use `AddingDefaultsBlockInputStream` (it's responsible for the defaults). I'll probably add it at some point.
For now you can just use https://clickhouse.tech/docs/en/sql_reference/functions/functions_for_nulls/#ifnull before inserting into target table

**Comment by alexey-milovidov** [comment]
@oandrew There is relevant issue #5213. Maybe you could help?

**Comment by oandrew** [comment]
@alexey-milovidov sure, I'll make a PR sometime soon.
